test/unit/instance-validator.test.ts

```diff
@@ -0,0 +1,551 @@
+import assert from 'node:assert';
+import { expect } from 'chai';
+import sinon from 'sinon';
+import type {
+  InferAttributes,
+  CreationOptional,
+  InferCreationAttributes,
+} from '@sequelize/core';
+import {
+  DataTypes,
+  ValidationError,
+  Model,
+  Sequelize,
+  literal,
+} from '@sequelize/core';
+// @ts-expect-error -- TODO: Migrate Instance Validator to TypeScript
+import { InstanceValidator } from '@sequelize/core/_non-semver-use-at-your-own-risk_/instance-validator.js';
+import {
+  getTestDialectTeaser,
+  sequelize,
+  beforeAll2,
+} from '../support';
+
+describe(getTestDialectTeaser('InstanceValidator'), () => {
+  const vars = beforeAll2(() => {
+    const User = sequelize.define('user', {
+      fails: {
+        type: DataTypes.BOOLEAN,
+        validate: {
+          isNotTrue(value: boolean) {
+            if (value) {
+              throw new Error('Manual model validation failure');
+            }
+          },
+        },
+      },
+    });
+
+    return { User };
+  });
+
+  it('allows setting an immutable field if the record is unsaved', async () => {
+    class User extends Model<InferAttributes<User>, InferCreationAttributes<User>> {
+      declare name: string;
+    }
+
+    User.init({
+      name: {
+        type: DataTypes.STRING,
+        validate: {
+          isImmutable: true,
+        },
+      },
+    }, { sequelize });
+
+    const user = User.build({ name: 'RedCat' });
+    expect(user.getDataValue('name')).to.equal('RedCat');
+
+    user.setDataValue('name', 'OrangeCat');
+    await expect(user.validate()).not.to.be.rejected;
+  });
+
+  it('raises an error for array on a STRING', async () => {
+    class User extends Model<InferAttributes<User>, InferCreationAttributes<User>> {
+      declare email: string;
+    }
+
+    User.init({
+      email: {
+        type: DataTypes.STRING,
+      },
+    }, { sequelize });
+
+    await expect(
+      User
+        // @ts-expect-error -- testing that this input will get rejected
+        .build({ email: ['ilove', 'sequelize.com'] })
+        .validate(),
+    ).to.be.rejectedWith(ValidationError);
+  });
+
+  it('raises an error for array on a STRING(20)', async () => {
+    class User extends Model<InferAttributes<User>, InferCreationAttributes<User>> {
+      declare email: string | string[];
+    }
+
+    User.init({
+      email: {
+        type: DataTypes.STRING(20),
+      },
+    }, { sequelize });
+
+    await expect(User.build({
+      email: ['ilove', 'sequelize.com'],
+    }).validate()).to.be.rejectedWith(ValidationError);
+  });
+
+  it('raises an error for array on a TEXT', async () => {
+    class User extends Model<InferAttributes<User>, InferCreationAttributes<User>> {
+      declare email: string | string[];
```

Comment:
The comment I wrote for the test above this one applies here too

```suggestion
      declare email: string;
```